### DISCUSSION TITLE
Should we limit smart casting?

### DISCUSSION THREAD

**Discussion by danieldaeschle** [vlang/v#7323] [General]
Kotlin has a limitation as well: https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS40LjIwIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsImpzQ29kZSI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiLyoqXG4gKiBZb3UgY2FuIGVkaXQsIHJ1biwgYW5kIHNoYXJlIHRoaXMgY29kZS4gXG4gKiBwbGF5LmtvdGxpbmxhbmcub3JnIFxuICovXG5cbmRhdGEgY2xhc3MgVGVzdCAodmFsIHg6IEFueSlcblxuZnVuIG1haW4oKSB7XG4gICAgdmFyIHQgPSBUZXN0KFRlc3QoXCJhc2RzYWRcIikpXG4gICAgaWYgKHQueCBpcyBUZXN0KSB7XG4gICAgICAgIHZhbCBhID0gdC54LnhcbiAgICB9XG59In0=

Just encountered this and want to share.

**Comment by Delta456** [comment]
That happens because `Any` is a sum type or it's like `interface{}` of Go? V compiler already uses nested sum types so I think we can have an exception for the compiler but I feel it shouldn't be limiting at all.

**Comment by medvednikov** [comment]
@danieldaeschle I think it's a serious limitation in Kotlin, and it's good that we don't have it.

We do not allow `if x.vals[0] is Type {` smart casting, I think that's reasonable.

**Comment by ntrel** [comment]
It's not a limitation of kotlin, it's because mutable smart casting can violate memory safety. This works:
```kotlin
    val t = Test(Test("asdsad"))
    if (t.x is Test) {
        val a = t.x.x
        println(a)
    }
```
